src/vm_pmu.rs

```diff
@@ -365,7 +368,111 @@ impl VmPmuState {
         }
     }
 
-    /// Calls the SBI configure_matching_counters() and performs internal bookkeeping on counter state.
+    fn configure_delegated_matching_counters(
+        &mut self,
+        counter_index: u64,
+        counter_mask: u64,
+        config_flags: PmuCounterConfigFlags,
+        event_type: PmuEventType,
+        event_data: u64,
+    ) -> SbiResult<u64> {
+        use sbi::PmuHardware::*;
+        fn configure_counter_csrs(
+            config_flags: PmuCounterConfigFlags,
+            event_type: PmuEventType,
+            event_data: u64,
+            counter_index: u64,
+        ) {
+            let mut shpmevent = LocalRegisterCopy::<u64, shpmevent::Register>::new(0);
+            // Translate uinh/sinh
+            if config_flags.is_uinh() {
+                shpmevent.modify(shpmevent::vuinh.val(1));
+            }
+            if config_flags.is_sinh() {
+                shpmevent.modify(shpmevent::vsinh.val(1));
+            }
+            shpmevent.modify(shpmevent::sinh.val(1));
+            // TODO: The mapping of events to platform counters is still under discussion.
+            // This will need to be revisited after the standard has been finalized.
+            match event_type {
+                PmuEventType::Hardware(_) | PmuEventType::Cache(_) => {
+                    shpmevent.modify(shpmevent::event_selector.val(event_type.raw()))
+                }
+                PmuEventType::RawEvent => {
+                    shpmevent.modify(shpmevent::event_selector.val(event_data))
+                }
+                _ => unreachable!(),
+            }
+            CSR.scounterselect.set(counter_index);
+            if !config_flags.is_skip_match() {
+                CSR.shpmevent.set(shpmevent.get());
+            }
+            if config_flags.is_clear_value() {
+                CSR.shpmcounter.set(0);
+            }
+            if config_flags.is_auto_start() {
+                CSR.scountinhibit.read_and_clear_bits(1 << counter_index);
+            }
+        }
+
+        let pmu_info = pmu::PmuInfo::get()?;
+        let counter_mask = pmu_info.filter_counter_mask(counter_index, counter_mask)?;
+        let index = if !config_flags.is_skip_match() {
```

Comment:
The logic in this block can be a bit hard to follow. I think it might be more readable if written like this:

```
let index = if config_flags.is_skip_match() {
    self.counter_state
        .get(counter_index as usize)
        .filter(|s| matches!(s, Started(_) | Configured(_)))
        .ok_or(SbiError::InvalidParam)
} else {
    let fixed_index = match event_type {
        PmuEventType::Hardware(CpuCycles) => Some(0),
        PmuEventType::Hardware(Instructions) => Some(2),
        PmuEventType::Firmware(_) => {
            return Err(SbiError::NotSupported);
        }
        _ => None,
    };
    let mut iter = CounterMaskIter::new(counter_index, counter_mask);
    iter.filter(|i| fixed_index.is_none() || fixed_index == Some(i))
        .find(|i| matches!(self.counter_state[i], NotConfigured))
        .ok_or(SbiError::NotSupported)
}?;
configure_counter_csrs(...);
```

Reply:
Nice use of of the Some / None trick. Also, good use of the non-iterator `.filter()` -- I likely read about it in the passing, but this is a great use case.